Add script to query app liveness for canary

maggie-lou opened the pull request:
<!--
Optional: Provide additional description (beyond the PR title).
Description should provide any background or motivation needed for the change, as well
as a high-level overview of the approach taken (if the change is not straightforward).
Detailed rationale for specific sections of the code are probably better off as code comments
so that future readers of that code can benefit.
-->

---

**Version bump**: TODO <!-- Required. Choose from: Major, Minor, Patch, None -->

<!-- See https://semver.org/#semantic-versioning-specification-semver. Summary:
* Major: Breaking change that causes existing functionality to not work as expected.
* Minor: Non-breaking change that adds functionality (examples: new feature; new API options)
* Patch: Non-breaking change that fixes an issue, improves performance, or refactors
         code.
* None:  Changed files are not included in releases (tests, docs, development setup,
         production configs)
-->

<!-- Optional:
**Related issues**: Fixes #1, Unblocks #2 ...
-->

maggie-lou commented:
Current dependencies on/for this PR:
* master
  * **PR #3318** <a href="https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3318" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/buildbuddy-io/buildbuddy/3318?utm_source=stack-comment).

tylerwilliams commented:
mentioned on slack but I think we should find a more standard way to specify the timeouts -- i sent you some options but lmk if you want to chat about this

maggie-lou commented on enterprise/tools/check_metrics/check_metrics.go:
```diff
@@ -0,0 +1,158 @@
+	}
+	promAPI := v1.NewAPI(prometheusClient)
+
+	eg := &errgroup.Group{}
```
Done

maggie-lou commented on enterprise/tools/check_metrics/check_metrics.go:
```diff
@@ -0,0 +1,158 @@
+					if sustainedUnhealthy {
+						log.Warningf("Metrics unhealthy for %s. Exiting script.", metric.Name)
+						os.Exit(metricsUnhealthyExitCode)
+					}
```
Done

maggie-lou commented on enterprise/tools/check_metrics/check_metrics.go:
```diff
@@ -0,0 +1,158 @@
+		})
+	}
+
+	eg.Wait()
```
Done

maggie-lou commented:
> mentioned on slack but I think we should find a more standard way to specify the timeouts -- i sent you some options but lmk if you want to chat about this

Refactored to use util.Retry -  let me know if that addresses your concerns

tylerwilliams commented on enterprise/tools/check_metrics/check_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		Address: config.PrometheusAddress,
+	})
+	if err != nil {
+		log.Fatalf("Error creating prometheus client: %v\nDid you run the prometheus port-forward command?", err)
```
nit: you can format errors with "%s", doesn't need to be %v

maggie-lou replied:
Done

tylerwilliams commented on enterprise/tools/check_metrics/check_metrics.go:
```diff
@@ -0,0 +1,134 @@
+	promAPI := v1.NewAPI(prometheusClient)
+
+	eg, gctx := errgroup.WithContext(context.Background())
+	for _, metric := range config.PrometheusMetrics {
```
nice; this is much simpler!

tylerwilliams reviewed (changes requested):
LGTM with a few small tweaks

maggie-lou commented on enterprise/tools/check_metrics/check_metrics.go:
```diff
@@ -0,0 +1,134 @@
+// Returns whether the given metric is healthy for the canary
+// If the canary's success rate is lower than that of the other apps by the input threshold, this will return false
+// Otherwise will return true
+func canaryHealthy(v1api v1.API, metricName string, canaryQuery string, baselineQuery string, canaryHealthThreshold float64) (bool, error) {
```
Done

bduffany commented on enterprise/tools/check_metrics/check_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		baselineMetricValue = baselineResultVector[0].Value
+	}
+
+	// If the metric differs from the baseline by the set threshold, the metric is considered unhealthy
```
since we're doing Abs() here could we get into a weird state where the canary is actually more healthy than the other apps and we're reporting it as unhealthy?

maggie-lou replied:
Yeah... I thought about that too. I guess that could happen if we fix an error, so the canary has fewer GRPC errors than the other apps for example

Tried to clarify this, but let me know if I just made it more confusing

maggie-lou replied:
Refactored the way to specify health thresholds. Hopefully it's a lot more clearer now

bduffany commented on enterprise/tools/check_metrics/check_metrics.go:
```diff
@@ -0,0 +1,134 @@
+		return false, err
+	}
+	resultVector := result.(model.Vector)
+	metricValue := model.SampleValue(0)
```
sanity check, does the 0'th value correspond to the latest value?

maggie-lou replied:
The API only returns one value for the current time. If there's data, the vector has size 1. Otherwise it's empty. This line `model.SampleValue(0)` is in case no data is returned, which happens for some of our counter metrics

maggie-lou commented on enterprise/tools/check_metrics/check_metrics.go:
```diff
@@ -0,0 +1,134 @@
+						unhealthyCount++
+					}
+
+					if unhealthyCount >= metric.MaxUnhealthyCount {
```
Personally representing it as a count is more intuitive to me than a percentage. I don't necessarily tie the two values together. For example, if the liveness check is down for 5 minutes, that is a clear signal to me we need a rollback because 5 minutes is a long time for an app to be down, not because that's 50% of the total duration I want to monitor the metrics. 

If it sounds good to you we can leave it as is for now and refactor if people are finding it confusing to configure?

maggie-lou replied:
Actually would it be more clear if it was MaxUnhealthyTimeframeSeconds instead of MaxUnhealthyCount?

maggie-lou commented on enterprise/tools/check_metrics/check_metrics.go:
```diff
@@ -0,0 +1,134 @@
+	"github.com/prometheus/common/model"
+	"golang.org/x/sync/errgroup"
+
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
```
Done
